### PR TITLE
[quant] Implement unsqueeze/squeeze for per-channel qtensor

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38407 [quant] Return default qconfig when backend is 'none'
* #38341 [quant][graphmode] Add quantized::conv1d to graphmode
* #38332 [quant] Add quantized::conv1d op benchmarck
* #38283 [quant] Add support for Quantized Conv1d and ConvRELU1d
* #38248 [quant] Add support for quantized::conv1d operator
* **#38247 [quant] Implement unsqueeze/squeeze for per-channel qtensor**

Summary:
Per-channel quantized tensor axis value is shifted based on the unsqueeze/squeeze dim

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_qtensor_unsqueze

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21550293](https://our.internmc.facebook.com/intern/diff/D21550293)

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -1232,6 +1239,32 @@ inferUnsqueezeGeometry(const Tensor& tensor, int64_t dim) {
...
+  }
+  if (self.dim() == 0 || self.sizes()[dim] != 1) {
+    sizes = self.sizes().vec();
+    strides = self.strides().vec();
+  }
+  auto result = make_qtensor(self, sizes, strides, quantizer);
+  namedinference::propagate_names_except(result, self, {dim});
+  return result;
+}
+
```
Comment (raghuramank100) [inline aten/src/ATen/native/TensorShape.cpp:1267]: Should we also add support for squeeze with no dimensions specified?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Looks good, a few suggestions to add more coverage

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
Comment (supriyar) [reply aten/src/ATen/native/TensorShape.cpp:1267]: Sure

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -613,6 +613,13 @@ Tensor sum_to_size(const Tensor& self, IntArrayRef size) {
   return sum_to(self, size);
 }
 
+Tensor make_qtensor(const Tensor& self, IntArrayRef size, IntArrayRef stride, QuantizerPtr quantizer) {
```
Comment (jerryzh168) [inline aten/src/ATen/native/TensorShape.cpp:616]: I think you can defined a new `as_strided_tensorimpl` and remove the current `as_strided_tensorimpl`

**Review by raghuramank100 — APPROVED** [review]

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
please make `make_qtensor` a native function and remove `as_strided_qtensorimpl`

**Comment by supriyar** [comment]
> please make `make_qtensor` a native function and remove `as_strided_qtensorimpl`

@jerryzh168 `as_strided_qtensorimpl` is still used by other ops like `unfold`, `diagonal`,  `expand`, `permute`. We'd have to update all those ops as well for per-channel quant. I think we can keep it as is for now and update later when we want to make these ops work for per-channel as well. Sounds good?

**Comment by jerryzh168** [comment]
Sure, but then could you create an issue, or add a TODO to reflect the plan?

**Review by jerryzh168 — APPROVED** [review]
need follow up PR to change the API of as_strided_qtensorimpl

**Pull request closed without merging**